src/python/pants/engine/target.py

```diff
@@ -555,87 +564,81 @@ class UnimplementedField(PrimitiveField, metaclass=ABCMeta):
     """
 
     value: Optional[Any]
-
-    # TODO: maybe implement __hash__ on this so that this can always be used with the engine? For
-    # example, `return hash(repr(raw_value))`?
-    #
-    # Whether we keep this Field or not really depends on how we approach backward-compatibility
-    # with V1. Are we indeed going to remove TargetAdaptor and use this API, which requires an
-    # explicit definition for each Field? How would this interact with custom target types? If we
-    # keep TargetAdaptor around for the sake of V1, then we should remove this Field.
-    def hydrate(self, raw_value: Optional[Any], *, address: Address) -> ImmutableValue:
-        return raw_value
+    default = None
 
 
 class BoolField(PrimitiveField, metaclass=ABCMeta):
     value: bool
     default: ClassVar[bool]
 
     def hydrate(self, raw_value: Optional[bool], *, address: Address) -> bool:
-        if raw_value is None:
-            return self.default
-        if not isinstance(raw_value, bool):
+        hydrated_value = super().hydrate(raw_value, address=address)
```

Comment:
I had no idea what to call this. It's not really `hydrated`..really, it's `the raw value or the default if the raw value was None`. 

Maybe `raw_value_or_default`?

Reply:
Maybe rename the method `calculate_value` or `compute_value` and let other names flow from there. Also, its odd  and potentially dangerous to have this be an instance method. Are there technical reasons preventing it from being a class method?

Reply:
For `PrimitiveField`, it can be a classmethod. 

For `AsyncField`, I intentionally had it be an instance method so that we could access `self.address`.

https://github.com/pantsbuild/pants/blob/23c4bf6fa06aa0330fdd2b0c893fcbecacddca2c/src/python/pants/engine/target.py#L189-L195

https://github.com/pantsbuild/pants/blob/23c4bf6fa06aa0330fdd2b0c893fcbecacddca2c/src/python/pants/engine/target.py#L702-L714

But that is probably not worth the reduction of boilerplate and complexity that we care about the order of `__init__` field initialization...